### ISSUE TITLE
twisted.names.dns._EDNSMessage should use the same repr mechanism as dns.Message

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6975]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6975|
|Type|enhancement|
|Created|2014-02-17 18:05:16Z|
|Branch|https://github.com/twisted/twisted/tree/concise-ednsmessage-repr-6975|

In ticket:6847#comment:8 suggested only showing non-default attribute values in the dns.Message repr.

It works really well, so I now want to apply the same mechanism to dns._EDNSMessage.

I'll refactor the code so that this "meaningful yet succinct" repr code can be shared and can be applied to other dns classes eg RRHeader and _OPTHeader.

See [twisted/names/dns.py#L2081](twisted/names/dns.py#L2081)

<details><summary>Searchable metadata</summary>

```
trac-id__6975 6975
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__EDNS0 EDNS0
branch__branches_concise_ednsmessage_repr_6975 branches/concise-ednsmessage-repr-6975
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1392660316000000 1392660316000000
changetime__1392897278000000 1392897278000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [41692]) Branching to 'concise-ednsmessage-repr-6975'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj removed owner|
|-|-|

Ready for review in log:branches/concise-ednsmessage-repr-6975
 * Refactored the dns.Message.__repr__ code so that it can be shared and re-used in _EDNSMessage.__repr__
 * Added an alwaysShow option to make the function flexible enough for use with classes other than dns messages.
 * Duplicated the dns.Message.__repr__ tests for _EDNSMessage.__repr__
 * Added dedicated tests for the new _compactRepr function.

In future, this new function could also be used in dns.RRHeader, dns._OPTHeader, dns._OPTVariableOption and maybe even the dns.Record_XXX classes. In those cases, "sectionNames" makes no sense, but those can be done in separate tickets.

Build Results:
 *

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

Build results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/concise-ednsmessage-repr-6975

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @wallrj|
|-|-|

Very nice.  Not a whole lot to comment on here; just a better implementation.  Please merge :).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set status to `closed`|
|-|-|

(In [41737]) Merge concise-ednsmessage-repr-6975

Author: rwall
Reviewer: glyph
Fixes: [#6975](https://github.com/twisted/twisted/issues/6975)

dns._EDNSMessage.__repr__ now only shows non-default attribute values making it
more concise and easier to read.
